backends/gstreamer/src/webrtc.rs

```diff
@@ -146,14 +177,128 @@ impl GStreamerWebRtcController {
+                // the caps are application/x-unknown by default, which will fail
+                // to intersect
+                //
+                // see https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/blob/ba62917fbfd98ea76d4e066a6f18b4a14b847362/ext/webrtc/gstwebrtcbin.c#L2521
+                caps_mut.get_mut_structure(cap).unwrap().set_name("application/x-rtp")
+            }
+            self.remote_mline_info.push(MLineInfo {
+                caps: caps,
+                // XXXManishearth in the (yet unsupported) case of dynamic stream addition and renegotiation
+                // this will need to be checked against the current set of streams
+                is_used: false,
+                // XXXManishearth ideally, we keep track of all payloads and have the capability of picking
+                // the appropriate decoder. For this, a bunch of the streams code will have to be moved into
+                // a webrtc-specific abstraction.
+                payload: media.get_format(0).unwrap().parse().unwrap(),
+            });
+        }
+    }
+
+    /// Streams need to be linked to the correct pads, so we buffer them up until we know enough
+    /// to do this.
+    ///
+    /// When we get a remote offer, we store the relevant m-line information so that we can
+    /// pick the correct sink pad and payload. Shortly after we look for any pending streams
+    /// and connect them to available compatible m-lines using link_stream.
+    ///
+    /// When we create an offer, we're controlling the pad order, so we set request_new_pads
+    /// to true and forcefully link all pending streams before generating the offer.
+    fn link_stream(&mut self, mut boxed_stream: Box<MediaStream>, request_new_pads: bool) {
+        let stream = boxed_stream
+            .as_mut_any()
+            .downcast_mut::<GStreamerMediaStream>()
+            .unwrap();
+        let caps = stream.caps();
+        let webrtc = self.webrtc.as_ref().unwrap();
+        let idx = self.remote_mline_info.iter().enumerate()
+            .filter(|(_, x)| !x.is_used)
+            .find(|(_, x)| x.caps.can_intersect(&caps))
+            .map(|x| x.0);
+        let element = stream.src_element();
+
+        if let Some(idx) = idx {
+            if idx >= self.request_pad_counter {
+                for _ in self.request_pad_counter..=idx {
+                    // webrtcbin needs you to request pads (or use element.link(webrtcbin))
+                    // however, it also wants them to be connected in the correct order.
+                    //
+                    // Here, we make sure all the numbered sink pads have been created beforehand, up to
+                    // and including the one we need here.
+                    //
+                    // An alternate fix is to sort pending_streams according to the m-line index
+                    // and just do it in order. This also seems brittle.
+                    webrtc.get_request_pad("sink_%u").unwrap();
```

Comment:
I feel like there's an underlying bug here, this should be unnecessary

cc @slomo